### DISCUSSION TITLE
The "magick mogrify -format" command changes the contrast (or gamma)

### DISCUSSION THREAD

**Discussion by MishaPhysics** [ImageMagick/ImageMagick#7377] [Help]
### ImageMagick version

7.1.1-33

### Operating system, version and so on

Windows 10

### Description

I convert the "file.jpg" picture to the *.dds format using:
`magick mogrify -format dds *.jpg`
Also, I convert the same "file.jpg" picture to the *.dds format using Photoshop. The problem is that the two output *.dds files are different. It seems that the ImageMagick changes the contrast (or gamma) during converting. Maybe there's some special options for the `magick mogrify -format` command?

I attach three files: file.jpg, the *.dds picture obtained with ImageMagick, and the *.dds picture obtained with Photoshop.
[Pictures.zip](https://github.com/user-attachments/files/15687243/Pictures.zip)

**Comment by fmw42** [comment]
I am not a DDS expert. But perhaps you need to use the -defines for DDS.  See https://imagemagick.org/script/defines.php

**Comment by MishaPhysics** [comment]
Thanks. I used `magick identify -verbose *.dds` to see the difference bethween two *.dds pictures. Some values are different, for example, these ones:
![c](https://github.com/ImageMagick/ImageMagick/assets/104359915/7e8f9efe-b6ac-456e-8c44-d8cd8fad95ec)
Also, I tried to use `-define` for the *.dds format (there are four commands), but I didn't understood where I need to put that command.

**Comment by snibgo** [comment]
> It seems that the ImageMagick changes the contrast (or gamma) during converting.

What makes you think this?

When we convert an image with two different software systems, we don't expect the results to be exactly the same.

When I open the two DDS files in Gimp, I can't see a visible difference.

The verbose info shows mimimum and maximum values are slightly different. These are outliers by definition, so are likely to be different.

The standard deviations are different, especially in the blue channel: ImageMagick.dds has 0.00352145, and photoshop.dds has 0.00312192. This difference is a large proportion of the value. However, the values are very low; the contrast in the blue channel is virtually zero. Hence, a small absolute difference gives a large proportional difference.

When I use Gimp to read file.jpg and export to g.dds, with compression=none, then read that with IM, the blue channel standard deviation is 0.00345924. So that is close to the ImageMagick.dds value of 0.00352145.

Conclusion: I can't see a problem.

**Reply by MishaPhysics** [reply]
Thanks for your reply.
I agree, the difference isn't noticeable if we just look at those two *.dds files. The thing is that I use these *.dds pictures as the custom texture tiles for a free space simulator. So, the difference is quite noticeable.
ImageMagick (you may notice a rectangle of a different tone):
![1-](https://github.com/ImageMagick/ImageMagick/assets/104359915/4018db68-6797-469c-8ed1-3a7a7d283b91)
Photoshop (the ocean has the same color):
![2-](https://github.com/ImageMagick/ImageMagick/assets/104359915/c889aa58-0816-4523-a1d7-cfbef9ef908a)
I really don't know why the converting in Photoshop gives such ideal result (color matching). Maybe I should think about using some action in Photoshop for batch converting of a large number of images.

**Comment by snibgo** [comment]
In the large blue area, `ImageMagick.dds` is a constant solid blue. `photoshop.dds` is also a constant solid blue, and the same as `ImageMagick.dds`.

In your last two images, neither has a constant solid blue. They are also different from each other. So you have done some processing to make those image that has introduced variation into solid colours, and the variation is different in the two images.

Naturally, this has also made differences in the "island" at top-right.

I don't know what processing has been done, so can't comment on that.

You last two images are JPG, so have been lossy-compressed. This is unlikely to affect the large solid blue area, but as a general rule, don't save images as JPG and then compare the results. You won't know if differences are due to the JPG compression or some other cause.

Your input to the OP DDS files is also JPG, `file.jpg`. This is a bad idea: JPG images contain lossy artifacts, and should never be used as input to image processing.

**Reply by MishaPhysics** [reply]
Yes, there's some effect in game (I don't know which).
I've tried to convert the *.png and *.bmp images to the *.dds format, but the result still differs from the one given by Photoshop.

**Reply by fmw42** [reply]
Might there be some auto brightness/contrast done in photoshop that IM does not do?

**Reply by MishaPhysics** [reply]
Maybe. Honestly, I'm not sure, but I'll keep trying.

**Reply by fmw42** [reply]
Honestly I see no differences.  I can display the two images and flicker them back and forth and nothing visually changes.  What is so different in your opinion and mathematically do they differ enough to make a difference.

I can do a pixel by pixel compare and get

magick compare -metric rmse 337684162-4018db68-6797-469c-8ed1-3a7a7d283b91.jpg 337684233-c889aa58-0816-4523-a1d7-cfbef9ef908a.jpg null:
208.024 (0.00317424)

This is a 0.3% difference on average.  That is hardly enough to make a difference and could be due to numerical precision or round-off.

**Comment by fmw42** [comment]
When you use Photoshop are you using the same color profiles? See the color settings under the edit menu. Imagemagick does not know anything about dot gain for color or gray. So be sure to disable those and set the color profile to the same as you use in Imagemagick

**Reply by MishaPhysics** [reply]
When I open my *.jpg picture in Photoshop I use the "Adobe RGB (1998)" profile by default. I looked at my "Color Settings" in Photoshop, there are - RGB: Adobe RGB (1988); Gray: Dot Gain 25%; Spot: Dot Gain 25%. I think the Photoshop color profile could really affects the converting result, since the input *.jpg file I'm using for converting I get from the Photoshop (I edit the input *.jpg picture in Photoshop and then convert it to *.dds using both Photoshop and ImageMagick, and the *.dds file I get from Photoshop has ideal color matching with the input *.jpg file, whereas ImageMagick - not).

**Reply by fmw42** [reply]
Adobe RGB is not the same as sRGB, which may be the default for Imagemagick.  You can use Adobe RGB with profiles in Imagemagick if you need to change it.

Why do you say PS is ideal?  How do you know which one is correct?

Did you turn off Dot Gain to see how the PS result compares to IM, since IM does not use that.  Also be sure you are using the same color profile in both PS and Imagemagick. Use either Adobe RGB or sRGB, but the same in both PS and IM.

**Reply by MishaPhysics** [reply]
I just look at the two last pictures above. On the first picture you can see a blue rectangle which is a little darker than the ocean around, whereas this rectangle is invisible on the second picture.

I haven't turn off Dot Gain yet (then I can get a different color, but I need the one that Photoshop gives now). But the thing is that I need to convert the *.jpg (which I get from Photoshop) to *.dds, and get the color result like the Photoshop converting gives.

Maybe I should to set the ImageMagick profile like in Photoshop, but I don't know how to do that.

**Comment by snibgo** [comment]
> I just look at the two last pictures above. On the first picture you can see a blue rectangle which is a little darker than the ocean around, whereas this rectangle is invisible on the second picture.

I can't see a darker blue rectangle in the first image.

You haven't said how you made those images, or supplied the input images to that process, so I can't reproduce the problem, so I can't comment further.

**Reply by MishaPhysics** [reply]
Here's a border of the rectangle I mean:
![1-](https://github.com/ImageMagick/ImageMagick/assets/104359915/6b99efe9-649f-465f-bff8-c7771cac43ae)

Sorry. I've made those two images as screenshots from a game.

**Reply by fmw42** [reply]
It does us no good to look at screen shots. We need to see your original input and the original outputs from PS and IM (before you go to DDS).  Even if you see a darker area, how do you know that is not the more correct images.  What are you comparing to?

Please answer all questions when asked. You keep putting off answering us.

**Comment by MishaPhysics** [comment]
I'm really sorry for confusing you. I'm ready to provide the all needed input and output images and describe what I do in detail.

1) I have an input picture "file.jpg". I attach it in archive.
2) I convert this picture with IM using the command `magick mogrify -format dds *.jpg`. I get the output image "file.dds". I attach it in archive (I renamed it to be "IM.dds").
3) I open the input "file.jpg" with Photoshop. During opening I get the window "Missing Profile". It's chosen "Leave as is (don't color manage)" by default. I press "OK". The picture is opened. I go "File" -> "Save As" -> I choose DDS format and image name like "file" -> I press "Save" -> In DDS Options I set: Format: DXT1, Mipmap: No, Alpha Channel: None -> I press "OK". So, the picture is saved, namely I get "file.dds". I attach it in archive (I renamed it to be "PS.dds").

The "IM.dds" and "PS.dds" are different. By the way they have different size. "IM.dds" is 170 KB, whereas "PS.dds" is 128 KB. Then I look at these two *.dds images in game, and they look different. I don't know why, but "PS.dds" looks noticeably better, whereas "IM.dds" looks darker compared to the background (I mean the rectangle which is visible in the first picture, I made its border as well). All what I would like to get is converting "file.jpg" from *.jpg to *.dds using IM command like `magick mogrify -format dds *.jpg` in Windows CMD. Maybe I should add some additional options regarding colorspace or something else to this command. For example, I found `magick myimage.jpg -colorspace RGB -resize 200% -colorspace sRGB mybigimage.jpg` command on IM website.

The archive with input "file.jpg", output "IM.dds" and output "PS.dds" is attached below.
[images.zip](https://github.com/user-attachments/files/15749921/images.zip)
I'm afraind I don't understand I don't know what other information to provide. I'm really  ashamed of my ignorance.

Also, I exported my "file.jpg" to *.dds format using GIMP. And it looks in game the same as "PS.dds", and it has the size 128 KB, namely the same as after conversion in PS. But I'm trying to get this result in IM, because I have a good script for IM.

**Reply by fmw42** [reply]
What game?  How is game changing the images for its own usage?  Do the images look different in some other viewer (Irfanview for example)?  How do they compare when using magick compare -metric rms.  See my comments above that they are insignificantly different on average (0.3% different).  How do they compare to your original JPG?  Why do you think the PS result is better?  Is it closer to the original or just your opinion of what you like?

**Reply by fmw42** [reply]
The images you show in your original post are much brighter than the ones in your zip folder. How did you adjust them?  Were they adjusted by you or by the game?  Were they adjusted exactly the same?

**Reply by fmw42** [reply]
Doing compares on your images, I get

```
magick compare -metric rmse IM.dds[0] PS.dds[0] null:
123.163 (0.00187935)

magick compare -metric rmse file.jpg IM.dds[0] null:
185.943 (0.0028373)

magick compare -metric rmse file.jpg PS.dds[0] null:
172.173 (0.00262719) 

``` 

What this shows is that your two DDS files differ by only 1.9%.  Your jpg and IM DDS differ by 2.8% and your jpg and PS DDS differ by 2.6%



These are insignificant spread over the whole image!

**Reply by fmw42** [reply]
I will try to repeat your processing trying to keep colorspace and dot gain in PS equivalent to IM and get back to you later.

**Reply by fmw42** [reply]
Can you try your processing with uncompressed DDS and without mipmaps, just the main level at the largest dimensions?  That will test if the issue is coming from either compression or mipmapping.

**Comment by MishaPhysics** [comment]
The game is "Orbiter Space Flight Simulator". It uses *.dds pictures like surface tiles for Earth.

The game changes these *.dds pictures for observations. I don't know how. I just try "IM.dds" and "PS.dds" images in turn and see the difference. The difference is only the noticeable darker rectangle. I have no other criteria for comparing "IM.dds" and "PS.dds" pictures. I see the difference only in game comparing the *.dds picture with neighbors textures. I didn't look at these two *.dds in other viewers. I can't compare output *.dds files with input *.jpg file, because the difference is noticeable only between "IM.dds" and "PS.dds" when they are used as textures in game.

Why do I think the PS result is better? Because "PS.dds" has much better color matching with background ocean and I don't see the darker rectangle as for "IM.dds". Еhis is an objective reason. Sorry, don't you see the darker rectangle I attached above? Even with dotted red border?

I'm sure the brightness was changed by game. I didn't change any game options, I just I replaced one picture ("IM.dds") with another ("PS.dds").

Also, the different size of "IM.dds" and "PS.dds" could say about something, I really don't know.

**Reply by MishaPhysics** [reply]
Also, I attach "GIMP.dds" below which I exported from "file.jpg" to *.dds using GIMP. I set BC1/DXT1 compression for this exporting. The "GIMP.dds" has the same size (128 KB) as "PS.dds", and looks visually the same as "PS.dds" in game, namely without darker rectangle.
[GIMP.zip](https://github.com/user-attachments/files/15750023/GIMP.zip)

**Reply by fmw42** [reply]
I see no differences in the original DDS images that you sent in your zip file.  I see only one image with the red to show where you are concerned, but I have nothing to compare to.

My PS does not seem to have an option to save as DDS?  How did you do that?  Was some other plugin needed?  How did you find that option?

**Reply by fmw42** [reply]
I am sorry, but I have no clue what PS and GIMP are doing differently.  And I don't seem to be able to export as DDS from my PS. So I don't know what else to do.

**Reply by MishaPhysics** [reply]
Yes, PS don't have DDS format by default. I installed a plugin which called "AdobeDDS_v0.6_win.zip". It includes "DDS.8bi" file.

Thanks for your help. It's OK. I'll try later something else.

**Reply by fmw42** [reply]
That plugin is only for Windows.  There is no Mac version.

**Comment by MishaPhysics** [comment]
There's so-called "Adobe DDS Plugin V 0.6 - Mac & Win" (in Google), but it's not necessary for trying. Honestly, I'm not sure it works with Mac (I haven't tried it).

**Reply by fmw42** [reply]
What do you mean "in Google"?  Mac's don't use the Google Play store.  They use Apple App Store.  I looked for a version of that plugin for the Mac, but NVIDIA only shows versions for Windows and GPU.

**Reply by fmw42** [reply]
OK I found that.  I will give it a try and see if the Mac version works.  Thanks.

**Reply by fmw42** [reply]
Sorry, that plugin does not seem to be recognized by PS 2024 on my Mac.

**Reply by MishaPhysics** [reply]
No problem. Thank you a lot!

**Reply by MishaPhysics** [reply]
I'm sorry to disturb you, but I managed to convert file.jpg to DDS format using IM (I experimented with this), and the output file.dds looks good in game (without a darker rectangle). I just replaced my old command `magick mogrify -format dds *.jpg` with the command `convert file.jpg -define dds:compression=none file.dds`. It works only with setting `-define dds:compression=none`, I tried all the options (dxt1 and dxt5) and they give the darker rectangle. The only one problem is that the output file.dds has a big size, namely 1 MB. I attach the new file.dds that has the size of 1 MB. Its name is "none.dds"
[images.zip](https://github.com/user-attachments/files/15754280/images.zip)

**Reply by fmw42** [reply]
You are on IM 7, so you should be using magick not convert.  Does this work? 

`magick file.jpg -define dds:compression=none file.dds`

Try 

`magick file.jpg -define dds:compression=DXTy file.dds`, where y=1,2 etc

Try each DXT compression to see if it works better in IM 7 using magick than magick mogrify, which is older technology.


I did ask you above to try without compression and also perhaps without mipmaps to see if that was acceptable. Glad that helped.  Perhaps the DXT compressions are not the same or there is a bug.

**Reply by MishaPhysics** [reply]
`magick file.jpg -define dds:compression=none file.dds` works and gives the same result as `convert file.jpg -define dds:compression=none file.dds`.

I tried `magick file.jpg -define dds:compression=DXTy file.dds` with y=1,2 and the results are the same as with `convert` instead of `magick`. There's a darker rectangle in game in these cases.

Since I need to convert all pictures in a directory I found a command `magick mogrify -format dds -define dds:compression=none *.jpg`. It works good. So the only `compression=none` gives the result I need, and the only one problem is a big size, namely 1 MB against of 128 KB which give Photoshop and Gimp conversion. Maybe I could use some command to reduce the size of the output *.dds file after conversion.

**Reply by fmw42** [reply]
You could try converting to pyramid TIFF with one of its compression formats.  But will your game accept that format? Otherwise look for another tool that simply compresses to DXTy.

**Reply by fmw42** [reply]
Check if libvips will convert to DDS.  You could also try scripting GIMP for batch processing.

**Reply by fmw42** [reply]
See 

https://convertio.co/vips-dds/ 
https://filestar.com/skills/vips/convert-vips-to-dds
https://www.vertopal.com/en/convert/vips-to-dds
https://sharkfoto.com/converter/vips-to-dds/#google_vignette

for an online converter to DDS

**Reply by fmw42** [reply]
What happens if you compress with DXTy, but do not use mipmaps, just the largest dimension image?  Will your game accept that and will it create the lower resolution layers of the mipmap for you.  Will the game compress your compression=none mipmap version for you?

**Reply by MishaPhysics** [reply]
Thanks.

I'm afraid the game requires the only *.dds pictures for its surface textures.

Maybe scripting GIMP for batch processing will be the solution, but I'm far from it yet (I never used GIMP scripts before). But it could be the good way, I'll try.

I think online converters can't help me, since I need to convert a lot of files, so I need a converter with scripts supported.

I tried to create an action in Photoshop in order to save all my *.jpg pictures in a directory as *.dds ones, but it didn't work (or I made something wrong).

Sadly, I don't know if the game creates MIP maps automatically. I suppose the game doesn't do that, but I'm not sure.

The interesting thing is that I choose DXT1 when converting my *.jpg to *.dds using Photoshop and GIMP, and they give the needed results.

**Reply by fmw42** [reply]
One of my links above looks like it will convert a group of images.  Check it again.

I do not know why IM compression is not the same.  Perhaps there is a bug. You could report that as a bug report on this forum.

The problem is it only shows up when the game brightens the images.  Statistically, there is insignificant differences on the average as I showed above. And it is not visually noticeable in the original dds images

**Comment by snibgo** [comment]
> Perhaps there is a bug. You could report that as a bug report on this forum.

But we don't have enough information to reproduce the problem.

The output image with the claimed "dark rectangle" is made from multiple inputs. (See the top-right "island" area.) We are told that when one of the inputs is from IM the problem occurs and the sea doesn't match; when it is from PS there is no problem. But what about the other required input(s)? Where did they come from?

**Reply by MishaPhysics** [reply]
Honestly, I'm not sure if it's a bug, maybe it's just a very exotic case. I just don't want to be misleading. So, I think you could know better if it should be posted "as bug" (at your discretion).
